File: Sesion 11/Ejercicio/Ejercicio3 S11/ejercicio3.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <boost/archive/text_oarchive.hpp>
#include <boost/archive/text_iarchive.hpp>

using namespace std;
using namespace boost::archive;

class Persona {
private:
    string nombre;
    int edad;

    // Implementación de serialización amigable con Boost.Serialization
    friend class boost::serialization::access;
    template <typename Archive>
    void serialize(Archive &ar, const unsigned int version) {
        ar & nombre;
        ar & edad;
    }

public:
    Persona() {}
    Persona(const string& nombre, int edad)
        : nombre(nombre), edad(edad) {}

    void setNombre(const string& nombre) { this->nombre = nombre; }
    string getNombre() const { return nombre; }

    void setEdad(int edad) { this->edad = edad; }
    int getEdad() const { return edad; }
};

// Función para serializar un objeto Persona en un archivo de texto
void serializarPersona(const Persona& persona, const string& nombreArchivo) {
    ofstream archivoSalida(nombreArchivo);
    if (archivoSalida.is_open()) {
        text_oarchive archivoTexto(archivoSalida);
        archivoTexto << persona;
        cout << "Persona serializada correctamente en " << nombreArchivo << endl;
    } else {
        cerr << "No se pudo abrir el archivo para escritura." << endl;
    }
}

// Función para deserializar un objeto Persona desde un archivo de texto
void deserializarPersona(const string& nombreArchivo) {
    Persona persona;
    ifstream archivoEntrada(nombreArchivo);
    if (archivoEntrada.is_open()) {
        text_iarchive archivoTexto(archivoEntrada);
        archivoTexto >> persona;
        cout << "Persona deserializada desde " << nombreArchivo << endl;
        cout << "Nombre: " << persona.getNombre() << endl;
        cout << "Edad: " << persona.getEdad() << endl;
    } else {
        cerr << "No se pudo abrir el archivo para lectura." << endl;
    }
}

int main() {
    // Crear una instancia de Persona
    Persona persona("Juan Perez", 30);

    // Serializar el objeto Persona en un archivo de texto
    serializarPersona(persona, "persona.txt");

    // Deserializar el objeto Persona desde el archivo de texto
    deserializarPersona("persona.txt");

    return 0;
}

File: Sesion 11/Ejercicio/Ejercicio3 S11/ejercicio3.py
import pickle

# Definición de la clase Persona
class Persona:
    def __init__(self, nombre, edad):
        self.nombre = nombre
        self.edad = edad

# Función para serializar un objeto Persona en un archivo
def serializar_persona(persona, nombre_archivo):
    try:
        with open(nombre_archivo, 'wb') as file:
            pickle.dump(persona, file)
        print(f"Persona serializada correctamente en {nombre_archivo}")
    except IOError as e:
        print(f"No se pudo abrir el archivo para escritura: {e}")

# Función para deserializar un objeto Persona desde un archivo
def deserializar_persona(nombre_archivo):
    try:
        with open(nombre_archivo, 'rb') as file:
            persona = pickle.load(file)
            print(f"Persona deserializada desde {nombre_archivo}")
            print(f"Nombre: {persona.nombre}")
            print(f"Edad: {persona.edad}")
    except IOError as e:
        print(f"No se pudo abrir el archivo para lectura: {e}")
    except pickle.PickleError as e:
        print(f"Error al deserializar la persona: {e}")

if __name__ == "__main__":
    # Crear una instancia de Persona
    persona = Persona("Juan Perez", 30)

    # Serializar el objeto Persona en un archivo
    serializar_persona(persona, "persona.pickle")

    # Deserializar el objeto Persona desde el archivo
    deserializar_persona("persona.pickle")
